modules/internal/ChapelSyncvar.chpl

```diff
@@ -37,285 +37,510 @@ for the state transition, one is non-deterministically selected to proceed and
     }
-    return ret;
-  }
 
-  /*
-     This method blocks until the sync variable is full. The state of the sync
-     variable remains full when this method completes.
+    /*
+      1) Block until the sync variable is full.
+      2) Read the value of the sync variable and leave the variable full
 
-     :returns: The value of the sync variable.
-  */
-  proc _syncvar.readFF() {
-    var ret: base_type;
-    on this {
-      var localRet: base_type;
-      chpl_rmem_consist_release();
-      __primitive("sync_wait_full_and_lock", this);
-      localRet = value;
-      __primitive("sync_mark_and_signal_full", this); // in case others are waiting
-      chpl_rmem_consist_acquire();
-      ret = localRet;
+      :returns: The value of the sync variable.
+    */
+    proc readFF() {
+      return syncVar.readFF();
     }
-    return ret;
-  }
 
-  /*
-     This method is non-blocking and the state of the sync variable is
-     unchanged when this method completes.
+    /*
+      1) Read the value of the sync variable
+      2) Do not inspect or change the full/empty state
 
-     :returns: The value of the sync variable.
-  */
-  proc _syncvar.readXX() {
-    var ret: base_type;
-    on this {
-      var localRet: base_type;
-      chpl_rmem_consist_release();
-      __primitive("sync_lock", this);
-      localRet = value;
-      __primitive("sync_unlock", this);
-      chpl_rmem_consist_acquire();
-      ret = localRet;
+      :returns: The value of the sync variable.
+    */
+    proc readXX() {
+      return syncVar.readXX();
     }
-    return ret;
-  }
 
-  /*
-     :arg val: New value of the sync variable.
+    /*
+      1) Block until the sync variable is empty.
+      2) Write the value of the sync variable and leave the variable full
+    */
+    proc writeEF(x : valType) {
+      syncVar.writeEF(x);
+    }
 
-     This method blocks until the sync variable is empty. The state of the sync
-     variable is set to full when this method completes. This method implements
-     the default write of a sync variable.
-  */
-  proc _syncvar.writeEF(val:base_type) {
-    on this {
-      chpl_rmem_consist_release();
-      __primitive("sync_wait_empty_and_lock", this);
-      value = val;
-      __primitive("sync_mark_and_signal_full", this);
-      chpl_rmem_consist_acquire();
+    /*
+      1) Block until the sync variable is full.
+      2) Write the value of the sync variable and leave the variable full
+    */
+    proc writeFF(x : valType) {
+      syncVar.writeFF(x);
+    }
+
+    /*
+      1) Write the value of the sync variable and leave the variable full
+    */
+    proc writeXF(x : valType) {
+      syncVar.writeXF(x);
+    }
+
+    // Do not allow implicit reads of sync/single vars.
+    pragma "no doc"
+    proc readThis(x) {
+      compilerError("sync variables cannot currently be read - use writeEF/writeFF instead");
+    }
+
+    // Do not allow implicit writes of sync/single vars.
+    pragma "no doc"
+    proc writeThis(x) {
```

Comment:
I can't comment on the exact line for some reason, but should we also update readThis and writeThis for "_singlevar" so that it only mentions "single" instead of "sync/single"?